File: scripts/test-paylines.py
"""Test payline configuration — run: python scripts/test-paylines.py"""

PAYLINES = {
    "h-top": ("Riga superiore", "horizontal", [(0, 0), (1, 0), (2, 0)]),
    "h-mid": ("Riga centrale", "horizontal", [(0, 1), (1, 1), (2, 1)]),
    "h-bot": ("Riga inferiore", "horizontal", [(0, 2), (1, 2), (2, 2)]),
    "v-left": ("Colonna sinistra", "vertical", [(0, 0), (0, 1), (0, 2)]),
    "v-mid": ("Colonna centrale", "vertical", [(1, 0), (1, 1), (1, 2)]),
    "v-right": ("Colonna destra", "vertical", [(2, 0), (2, 1), (2, 2)]),
    "d-tl-br": ("Diagonale TL-BR", "diagonal", [(0, 0), (1, 1), (2, 2)]),
    "d-tr-bl": ("Diagonale TR-BL", "diagonal", [(2, 0), (1, 1), (0, 2)]),
}

PAYLINES_BY_BET = {
    250: ["h-top", "h-mid", "h-bot"],
    500: ["h-top", "h-mid", "h-bot", "v-left", "v-mid", "v-right"],
    1000: ["h-top", "h-mid", "h-bot", "v-left", "v-mid", "v-right", "d-tl-br", "d-tr-bl"],
}

SAMPLE_GRID = [
    ["cherry", "cherry", "cherry"],
    ["bell", "cherry", "star"],
    ["cherry", "cherry", "cherry"],
]


def main() -> None:
    print("=== ROYAL SLOT — Test linee vincenti ===\n")

    for bet, line_ids in PAYLINES_BY_BET.items():
        print(f"PUNTATA {bet} CHIP — {len(line_ids)} linee controllate:")
        winners = []

        for index, line_id in enumerate(line_ids, start=1):
            name, line_type, coords = PAYLINES[line_id]
            symbols = [SAMPLE_GRID[reel][row] for reel, row in coords]
            is_win = symbols[0] == symbols[1] == symbols[2]
            status = f"VINCENTE ({symbols[0].upper()} x3)" if is_win else "nessuna combinazione"
            print(f"  {index}. [{line_type}] {name} ({line_id}) -> {status}")
            if is_win:
                winners.append(name)

        if winners:
            print(f"  -> Combinazioni rilevate: {', '.join(winners)}")
        else:
            print("  -> Combinazioni rilevate: nessuna")
        print()

    print("Griglia di test (rullo → righe top|mid|bot):")
    for reel_index, column in enumerate(SAMPLE_GRID, start=1):
        print(f"  Rullo {reel_index}: {' | '.join(column)}")


if __name__ == "__main__":
    main()

File: scripts/test-paylines.js
/**
 * Test payline configuration — run: node scripts/test-paylines.js
 */

import {
  PAYLINES_BY_BET,
  getEnabledPaylines,
  evaluateEnabledPaylines,
  calculateWinAmount,
} from '../js/paylines.js';

const sampleGrid = [
  ['cherry', 'cherry', 'cherry'],
  ['bell', 'cherry', 'star'],
  ['cherry', 'cherry', 'cherry'],
];

/** Griglia con vincita su una sola linea orizzontale (bet 250). */
const singleWinGrid = [
  ['bell', 'horseshoe', 'star'],
  ['cherry', 'horseshoe', 'bar'],
  ['seven', 'horseshoe', 'crown'],
];

/** Griglia con vincita orizzontale + verticale contemporanea (bet 500). */
const multiWinGrid = [
  ['bell', 'horseshoe', 'star'],
  ['horseshoe', 'horseshoe', 'horseshoe'],
  ['seven', 'horseshoe', 'crown'],
];

console.log('=== ROYAL SLOT — Test linee vincenti ===\n');

Object.entries(PAYLINES_BY_BET).forEach(([bet, lineIds]) => {
  const enabled = getEnabledPaylines(Number(bet));
  const evaluation = evaluateEnabledPaylines(sampleGrid, Number(bet));

  console.log(`PUNTATA ${bet} CHIP — ${enabled.length} linee controllate:`);
  enabled.forEach((line, index) => {
    const result = evaluation.results[index];
    const status = result.win
      ? `VINCENTE (${result.match.toUpperCase()} x3)`
      : 'nessuna combinazione';
    console.log(`  ${index + 1}. [${line.type}] ${line.name} (${line.id}) → ${status}`);
  });

  const winners = evaluation.winningLines.map((entry) => entry.line.name);
  const totalWin = calculateWinAmount(evaluation);
  console.log(`  → Combinazioni rilevate: ${winners.length ? winners.join(', ') : 'nessuna'}`);
  console.log(`  → Vincita totale: ${totalWin.toLocaleString('it-IT')} CHIP\n`);
});

console.log('=== Simulazione 1: una linea vincente (250 CHIP) ===');
const singleEval = evaluateEnabledPaylines(singleWinGrid, 250);
const singleWin = calculateWinAmount(singleEval);
console.log(`Linee: ${singleEval.winningLines.map((e) => e.line.name).join(', ') || 'nessuna'}`);
console.log(`Atteso: horseshoe x3 → round(5 × 250 × 1.655634) = 2.070 CHIP`);
console.log(`Calcolato: ${singleWin.toLocaleString('it-IT')} CHIP`);
console.log(singleWin === 2070 ? 'OK\n' : 'ERRORE\n');

console.log('=== Simulazione 2: più linee vincenti (500 CHIP) ===');
const multiEval = evaluateEnabledPaylines(multiWinGrid, 500);
const multiWin = calculateWinAmount(multiEval);
console.log(`Linee: ${multiEval.winningLines.map((e) => e.line.name).join(', ')}`);
console.log(`Atteso: 2 × round(5 × 500 × 0.827817) = 4.140 CHIP`);
console.log(`Calcolato: ${multiWin.toLocaleString('it-IT')} CHIP`);
console.log(multiWin === 4140 ? 'OK\n' : 'ERRORE\n');

console.log('Griglia di test (reels x righe):');
sampleGrid.forEach((column, reelIndex) => {
  console.log(`  Rullo ${reelIndex + 1}: ${column.join(' | ')}`);
});
